File: dart_lang/lib/behavioral/mediator/chat_management.dart
import 'package:dart_lang/behavioral/mediator/chat_mediator.dart';
import 'package:dart_lang/behavioral/mediator/mediator_user_model.dart';

class ChatManagmenet implements ChatMediator {
  Map<String, List<MediatorUserModel>> groups;
  ChatManagmenet(this.groups);

  @override
  void registerUserToGroup(String groupName, MediatorUserModel user) {
    if (groups.containsKey(groupName)) {
      groups[groupName]!.add(user);
      return;
    } else {
      groups[groupName] = [user];
      return;
    }
  }

  @override
  void sendDirectMessage(
      String message, MediatorUserModel toUser, MediatorUserModel fromUser) {
    print(
        'from user ${fromUser.name} Sending Message ${message} to user${toUser.name} ');
    toUser.reciveDirectMessage(message, fromUser);
  }

  @override
  void sendGroupMessage(
      String message, String groupName, MediatorUserModel fromUser) {
    print(
        'from  ${fromUser.name} Sending Message ${message} to group${groupName}');
    if (groups.containsKey(groupName)) {
      groups[groupName]!.forEach((element) {
        if (element != fromUser) {
          element.reciveMessageeGroup(message, groupName, fromUser);
          return;
        }
      });
      return;
    }
  }
}

File: typescript/patterns/mediator/chat-management.ts
class ChatManagement implements ChatMediator {
  private groups: Map<string, User[]>;
  constructor() {
    this.groups = new Map<string, User[]>();
  }

  sendDirectMessage(message: string, sender: User, receiver: User): void {
    console.log(`
          Sending new message
          from user: ${sender.getName()}
          to user: ${receiver.getName()}
          content: ${message} 
          `);
    receiver.receiveDirectMessage(message, sender);
  }

  sendGroupMessage(message: string, groupName: string, sender: User): void {
    console.log(`
          Sending new message
          from user: ${sender.getName()}
          to group: ${groupName}
          content: ${message}
          `);
    this.groups
      .get(groupName)
      ?.filter((user) => user !== sender)
      .forEach((user) => {
        user.receiveGroupMessage(message, groupName, sender);
      });
  }

  registerUserToGroup(user: User, groupName: string): void {
    const group = this.groups.get(groupName);
    if (!group) {
      this.groups.set(groupName, [user]);
      return;
    }

    group.push(user);
  }
}
